### PR TITLE
Data pipeline

### PR DESCRIPTION
**Pull request by nlubalo**
Review the new updated db_script.py.

### CODE REVIEW

File: src/clean_and_process.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import numpy as np
+import scipy
+import datetime
+from scipy import stats
+
+from pipeline_func.clean import clean
+from pipeline_func.process import process
+
+ENGINE = engine.get_engine()
+ENGINE_rename = engine.get_engine()
```
Comment (dsementsov) [inline src/clean_and_process.py]: You don't need that variable

File: src/clean_and_process.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+module_json = 'config/indicators.json'
+with open(module_json) as f:
+    for section in json.load(f):
+        for p in section['data']:
+            ENGINE[p['identifier']] = p['value']
+
```
Comment (dsementsov) [inline src/clean_and_process.py]: This is a duplication of the get_engine file

File: src/clean_and_process.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+###########################
```
Comment (dsementsov) [inline src/clean_and_process.py]: Can we have this file at the root of repo and rename it to app.py?

File: src/db_insert.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import psycopg2
+import pandas as pd
+import json 
+import sys
+
+def get_engine(json_file, sec):
+    with open(json_file) as f:
+        ENGINE = dict((p['identifier'], p['value']) for section in json.load(f) 
+                                                    for p in section[sec])
+        return ENGINE
```
Comment (dsementsov) [inline src/db_insert.py]: This is a duplicate of the get_engine file

File: src/pipeline_func/clean.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+ENGINE = engine.get_engine()
+ENGINE_rename = engine.get_engine()
+
+module_json = 'config/indicators.json'
+with open(module_json) as f:
+    for section in json.load(f):
+        for p in section['data']:
+            ENGINE[p['identifier']] = p['value']
+
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: this is a duplicate of the get_engine

File: src/pipeline_func/clean.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+startTime = datetime.datetime.now()
+
+
+def make_note(statement):
+    print(statement, str(datetime.datetime.now() - startTime))
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: This method is duplicated. Move it to separate helpers.py file and import when needed.

File: src/pipeline_func/clean.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+
+###################################
+#     Define functions used       #
+###################################
+
+
+# Get user-input data
+
+VAR_CORR_PATH = ENGINE['var_correspondance_data']
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: this variable is not needed. Just read from the key

File: src/pipeline_func/clean.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+
+    return df1
+
+# Selecting target indicators
+
+
+USECOLS = list(range(0, 10))
+DTYPES = {'Unnamed: 0': int, 'dataElement': str, 'period': str, 'orgUnit': str, 'categoryOptionCombo': str,
+          'attributeOptionCombo': str, 'value': object, 'storedBy': str, 'created': str, 'lastUpdated': str}
+
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: Constants should be at the top of the file

File: src/pipeline_func/clean.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+def get_data(path, instance):
+
+    data = pd.read_csv(path, usecols=USECOLS, dtype=DTYPES)
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: you can also set the index to be Unnamed: 0 right away

File: src/pipeline_func/clean.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    return df
+
+# Transforming dates
+
+
+def date_convert(date):
+    if date.find("W") == -1:
+        x = pd.to_datetime(date, format='%Y%m')
+    else:
+        y = date.split("W")[0]
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: y, w = (date.split('W'))

File: src/pipeline_func/clean.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    old_var_add_dict = get_variable_addition_dict(VAR_CORR, 'old')
+
+    # Reporting data
+
+    new_dhis_report_df = get_reporting_data(new_dhis_report_path, "new")
+    make_note('new reporting data formatted for use')
+
+    old_dhis_report_df = get_reporting_data(old_dhis_report_path, "old")
+    make_note('old reporting data formatted for use')
+
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: Allocate data resources only when you need them!

If you want this to be less RAM efficient, I would even read-write the data in chunks.

You are doing:

read data 1 
read data 2 
read data 3

process data 1
process data 2 
process data 3

You need to be doing:

read data 1
process data 1
return data 1 or write data 1

read data 2 
process data 2
...

File: src/pipeline_func/clean.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    # Old instance
+
+    old_dhis_df = get_data(old_dhis_path, "old")
+    for x in list(old_var_add_dict.keys()):
+        old_dhis_df = compute_indicators(old_dhis_df, x, old_var_add_dict[x])
+    make_note('old data additional indicators created')
+
+    old_dhis_df = process_date(old_dhis_df)
+    make_note('old data transformed and cleaned')
+
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: This is highly redundant and creates an overhead!

File: src/pipeline_func/clean.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    new_dhis_df['dataElement'].replace(renaming_dict, inplace=True)
+    old_dhis_report_df['dataElement'].replace(renaming_dict, inplace=True)
+    new_dhis_report_df['dataElement'].replace(renaming_dict, inplace=True)
+    make_note('variable names updated')
+
+    # concatenate old and new instance
+
+    combined_df = pd.concat([old_dhis_df, new_dhis_df,
+                             old_dhis_report_df,
+                             new_dhis_report_df])
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: I am not sure we need reporting and data in the same df though

File: src/pipeline_func/clean.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    # Add district columns
+
+    districts = pd.merge(combined_df['orgUnit'], old_dhis_report_df[[
+        'orgUnit', 'districts']], how='left', left_on='orgUnit', right_on='orgUnit')
+    combined_df['districts'] = districts['districts']
+    make_note('district column added')
+
+    # Selecting only the facilities that are in both old and new
+    old_ids = set(old_dhis_report_df['orgUnit'].unique())
+    new_ids = set(new_dhis_report_df['orgUnit'].unique())
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: set and unique() does the same thing

File: src/pipeline_func/clean.py [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+    list_ids = list(new_ids.intersection(old_ids))
```
Comment (dsementsov) [inline src/pipeline_func/clean.py]: now what you could do is just 

set(old_dhis_report_df['orgUnit'].tolist() + new_dhis_report_df['orgUnit'].tolist())

Comment (dsementsov) [reply src/pipeline_func/clean.py]: you also could have filtered the dfs before concatting them together

File: src/pipeline_func/process.py [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+import os
+from pathlib import Path
+import engine
+import numpy as np
+import scipy
+import datetime
+from scipy import stats
+
+ENGINE = engine.get_engine()
+ENGINE_rename = engine.get_engine()
```
Comment (dsementsov) [inline src/pipeline_func/process.py]: unused variable

File: src/pipeline_func/process.py [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+
+module_json = 'config/indicators.json'
+with open(module_json) as f:
+    for section in json.load(f):
+        for p in section['data']:
+            ENGINE[p['identifier']] = p['value']
+
```
Comment (dsementsov) [inline src/pipeline_func/process.py]: duplication of get_engine

File: src/pipeline_func/process.py [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+startTime = datetime.datetime.now()
+
+
+def make_note(statement):
+    print(statement, str(datetime.datetime.now() - startTime))
```
Comment (dsementsov) [inline src/pipeline_func/process.py]: duplicate method

File: src/pipeline_func/process.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+
+# Get user-input data
+
+
+VAR_CORR_PATH = ENGINE['var_correspondance_data']
```
Comment (dsementsov) [inline src/pipeline_func/process.py]: unnecessary declaration

File: src/pipeline_func/process.py [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+#############################
+
+def process(data):
+
+    # separe reports and non reports indicators
+    report_indics = list(VAR_CORR[VAR_CORR['domain'] == 'REPORT']
+                         ['identifier'].unique())
+    data_df_noreport = data[~data['dataElement'].isin(
+        report_indics)].copy()
+    data_df_report = data[data['dataElement'].isin(report_indics)].copy()
```
Comment (dsementsov) [inline src/pipeline_func/process.py]: you allocate memory 20 lines before you use it.

**Review by dsementsov — CHANGES_REQUESTED** [review]

**Comment by nbigaud** [comment]
Outdated I think

**Pull request closed without merging**